libraries/AC_Avoidance/AP_OADijkstra.h

```diff
@@ -0,0 +1,115 @@
+    bool update(const Location &current_loc, const Location &destination, Location& origin_new, Location& destination_new);
+
+private:
+
+    // check if polygon fence has been updated since we created the inner fence. returns true if changed
+    bool check_polygon_fence_updated() const;
+
+    // create a smaller polygon fence within the existing polygon fence
+    // returns true on success
+    bool create_polygon_fence_with_margin(float margin_cm);
+
+    // create a visibility graph of the polygon fence
+    // returns true on success
+    // requires create_polygon_fence_with_margin to have been run
+    bool create_polygon_fence_visgraph();
+
+    // calculate shortest path from origin to destination
+    // returns true on success
+    // requires create_polygon_fence_with_margin and create_polygon_fence_visgraph to have been run
+    // resulting path is stored in _shortest_path array as vector offsets from EKF origin
+    bool calc_shortest_path(const Location &origin, const Location &destination);
+
+    // shortest path state variables
+    bool _polyfence_with_margin_ok = false;
+    bool _polyfence_visgraph_ok = false;
+    bool _shortest_path_ok = false;
+
+    Location _destination_prev;     // destination of previous iterations (used to determine if path should be re-calculated)
+    uint8_t _path_idx_returned;     // index into _path array which gives location vehicle should be currently moving towards
+
+    // polygon fence (with margin) related variables
+    float _polyfence_margin = 10;
+    Vector2f _polyfence_pts[OA_DIJKSTRA_POLYGON_FENCE_PTS];
+    uint8_t _polyfence_numpoints;
+    uint32_t _polyfence_update_ms;  // system time of boundary update from AC_Fence (used to detect changes to polygon fence)
+
+    // visibility graphs
+    AP_OAVisGraph _polyfence_visgraph;
+    AP_OAVisGraph _source_visgraph;
+    AP_OAVisGraph _destination_visgraph;
+
+    // updates visibility graph for a given position which is an offset (in cm) from the ekf origin
+    // requires create_polygon_fence_with_margin to have been run
+    // returns true on success
+    bool update_visgraph(AP_OAVisGraph& visgraph, const AP_OAVisGraph::OAItemID& oaid, const Vector2f &position, bool add_extra_point = false, Vector2f extra_point = Vector2f(0,0));
+
+    typedef uint8_t node_index;         // indices into short path data
+    struct ShortPathNode {
+        AP_OAVisGraph::OAItemID id;     // unique id for node (combination of type and id number)
+        bool visited;                   // true if all this node's neighbour's distances have been updated
+        float distance_cm;              // distance from source (number is tentative until this node is the current node and/or visited = true)
+        node_index distance_from_idx;   // index into _short_path_data from where distance was updated (or 255 if not set)
+    } _short_path_data[OA_DIJKSTRA_POLYGON_SHORTPATH_PTS];
```

Comment:
ok txs.  it makes it less readible but for 4bytes per entry that's quite valuable. excuse my ignorance but how about if "id" was placed below "distance_from_idx" would the 4 bytes be saved then too?

Reply:
Structs (unless you tag them with `PACKED` (which is actually a macro for compiler attributes)), have all their members aligned on word access. So the id is 1 byte, that's fine. But the 4 byte float won't take 3 bytes from the first word and 1 from the second as that's slower. But this meant you had 3 bytes of padding that weren't doing anything for you.

You could tag this PACKED and keep the current structure, but it would slow down the runtime performance. Rearranging can't always get as tight as packed can, but it doesn't sacrifice runtime performance.

Reply:
ok, cool, txs

Reply:
ah, actually the id is two bytes.. so i'll re-order it like this:

>     struct ShortPathNode {
>         AP_OAVisGraph::OAItemID id;     // unique id for node (combination of type and id number)
>         bool visited;                   // true if all this node's neighbour's distances have been updated
>         node_index distance_from_idx;   // index into _short_path_data from where distance was updated (or 255 if not set)
>         float distance_cm;              // distance from source (number is tentative until this node is the current node and/or visited = true)
>     } _short_path_data[OA_DIJKSTRA_POLYGON_SHORTPATH_PTS];